### PR TITLE
Refactor primitive type fields to be of object type in JSON objects

### PR DESCRIPTION
**Pull request by marcuslinke**

### CODE REVIEW

File: src/main/java/com/github/dockerjava/api/async/ResultCallback.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package com.github.dockerjava.api.async;
 
 import java.io.Closeable;
+import java.io.IOException;
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/async/ResultCallback.java]: do you need it?

File: src/main/java/com/github/dockerjava/api/model/BuildResponseItem.java [hunk 2]
```diff
@@ -25,7 +25,7 @@ public String getStream() {
      * Returns whether the stream field indicates a successful build operation
      */
     @JsonIgnore
-    public boolean isBuildSuccessIndicated() {
+    public Boolean isBuildSuccessIndicated() {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/BuildResponseItem.java]: null can't be returned here

File: src/main/java/com/github/dockerjava/api/model/PullResponseItem.java [hunk 3]
```diff
@@ -17,7 +17,7 @@
      * @returns true: status indicates that pull was successful, false: status doesn't indicate a successful pull
      */
     @JsonIgnore
-    public boolean isPullSuccessIndicated() {
+    public Boolean isPullSuccessIndicated() {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/PullResponseItem.java]: keep primitive, it's not used in JSON and can't have null?

File: src/main/java/com/github/dockerjava/api/model/PushResponseItem.java [hunk 4]
```diff
@@ -17,7 +17,7 @@
      * @returns true: the error field indicates an error, false: the error field doesn't indicate an error
      */
     @JsonIgnore
-    public boolean isErrorIndicated() {
+    public Boolean isErrorIndicated() {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/PushResponseItem.java]: the same

File: src/main/java/com/github/dockerjava/core/async/ResultCallbackTemplate.java [hunk 5]
```diff
@@ -28,7 +28,7 @@
 
     private Closeable stream;
 
-    private boolean closed = false;
+    private Boolean closed = false;
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/async/ResultCallbackTemplate.java]: is it makes sense?

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java [hunk 6]
```diff
@@ -120,12 +118,12 @@ public Boolean getForcerm() {
     }
 
     @Override
-    public boolean isQuiet() {
+    public Boolean isQuiet() {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: AFAIK `is*` can be used only with primitives, after making it Object it should be `get*`( or `has*`)

File: src/test/java/com/github/dockerjava/core/command/ListImagesCmdImplTest.java [hunk 7]
```diff
@@ -72,11 +72,11 @@ public void listDanglingImages() throws DockerException {
         assertThat(images, notNullValue());
         LOG.info("Images List: {}", images);
         assertThat(images.size(), is(greaterThan(0)));
-        boolean imageInFilteredList = isImageInFilteredList(images, imageId);
+        Boolean imageInFilteredList = isImageInFilteredList(images, imageId);
         assertTrue(imageInFilteredList);
     }
 
-    private boolean isImageInFilteredList(List<Image> images, String expectedImageId) {
+    private Boolean isImageInFilteredList(List<Image> images, String expectedImageId) {
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/ListImagesCmdImplTest.java]: it always primitive, no sense in returning Object

**Comment by KostyaSha** [comment]
@marcuslinke my thoughts:
- `is*` should be used only for boolean primitives, `get*` or `has*` for Objects
- primitive -> Object doesn't make sense for stream handling/callback internals (was all this change done with sed?)
- all Nullable return values and fields should be annotated with `@CheckForNull`
- `withSomething()` looks redundant in interface and it not so difficult to set explicit values for library consumers
- many places will have null checks for data values `if (getSomething() != null && getSomething() {` so maybe it makes sense to keep `isSomething()` methods that will have this null check internally (see BuildImageCmd where i kept backward compatibility for primitive method and used Object for data field

**Comment by marcuslinke** [comment]
@KostyaSha Thanks for the review! Yes, I initially did search and replace for booleans, so I need to review all of these replacements. I like your suggestions especially that it will retain backward compatibility. However, it will take some time as I'm quite busy with other things at the moment.

**Comment by KostyaSha** [comment]
@marcuslinke my suggestion will be to start doing it partially. I can start with critical for docker-plugin API commands and PR changes. @marcuslinke btw, backward compatibility will be broken in any release because all methods defined in interface that requires extending fields/methods for every docker API update. So it seems that we should forget about compatibilities in X.Y.

**Comment by marcuslinke** [comment]
@KostyaSha After thinking about it again I want to discuss these additional get/has/is methods that return primitives (commands):
* Examining existing command objects isn't the general use case I guess, so handling null cases might be acceptable for the caller
* Returning primitives introduces some kind of default value which might be OK for boolean but what about long/int values (see `tail` member in log command for example)?

**Comment by KostyaSha** [comment]
Maybe attach to request only non-null values without checking what is default plus annotate everything with `@CheckForNull`? If docker has some default values, then it's probably not a docker-java issue to show user what is default.

**Comment by marcuslinke** [comment]
OK, so lets summarize:
* dropping  get/is/has methods that return primitives like `boolean`, `long`, `int`
* by annotating json objects with `@JsonInclude(Include.NON_NULL)` we can consider that only non-null values are sent with the request
* by annotating appropriate get/has/is methods with `@CheckForNull` FindBugs can find missing null checks in client code

Do you agree?

**Comment by KostyaSha** [comment]
Sounds good

**Comment by KostyaSha** [comment]
My idea about removing  `withSomething()` was to minimise code in library.

**Comment by marcuslinke** [comment]
Yeah. I will probably remove this also. By the way, according to http://ehc.ac/p/findbugs/bugs/334/ `@CheckForNull` can be annotated at the interface level. So I will remove this annotation from the implementation ('BuildImageCmdImpl'), OK?

**Comment by KostyaSha** [comment]
Right, for method return values inheritance should work.

**Comment by KostyaSha** [comment]
@marcuslinke and if we do refactoring, is it possible to use Lists instead arrays?

**Comment by marcuslinke** [comment]
@KostyaSha I think we could add additional methods that takes `List`s.

**Comment by KostyaSha** [comment]
Disclaimer: i'm not an expert of api libraries :)
@marcuslinke my main concern about List was to have easier data manipulation. For example in docker i fill create command on 2 levels. First is configuration and then launcher modifies parameters. With array if very inconvenient to do changes, it enforces casting to List, working with it and back setting to array. 
So logical idea was to simplify api and place List directly.

**Comment by KostyaSha** [comment]
According to google this code should work
```
@JsonProperty("Cmd")
private List<String> cmd;
```
Then you can simply use some http://commons.apache.org/proper/commons-collections/javadocs/api-3.2.1/org/apache/commons/collections/CollectionUtils.html#isNotEmpty(java.util.Collection) or guava...

**Comment by marcuslinke** [comment]
I think the main difference between arrays and lists is the aspect of mutablitity. Whereas arrays are immutable, lists in contast are not and allow the manipulation from the outside. In general I dont think this is a good idea as it may result in unexpected behaviour. So switching to lists would enforce to ensure the immutablitity by copying to (when implementing the setter) and copying from (when implementing the getter) the internal list. I would avoid this implementation effort...

**Comment by KostyaSha** [comment]
Where do you see mutability issue? You can either replace List either manipulate with existed. During exec it will be send to remote.

**Pull request merged**